Interpolation

hcmh opened the pull request:
This pull request does two things:

1) It corrects the interpolation used in view. Previously, we had an ugly black border since the step size was correct, but the start was not: We actually need to start at negative positions and set pixels outside of the valid range to their nearest neighbor.

2) It adds support for nearest neighbor interpolation, which can be chosen from the GUI and as a command-line option in`cfl2png`.


Here are two small test datasets (2x2 and 3x3) which can be used to verify the interpolation:
https://gist.github.com/hcmh/c9dde7d52bb322187bfc9c9d15bc89e8

uecker commented:
Excellent. Here are some minor comments:

- coding style (esp. spaces around operators).
- I am not sure computation is right for negative dx/dy. Please check.
- I would say error is for wrong user input. "Can't happen" should be assert(0)
- the computation of the negative shifts can be simplified a lot by expanding the fraction with dx (this even removes the if conditions)

uecker commented:
`....0.......1.....d....(d - 1.) / 2.`
`|---*---|---*---| 1.00 -> -0.000`
`|-*-|-*-|-*-|-*-| 0.50 -> -0.250`
`|*|*|*|*|*|*|*|*| 0.25 -> -0.375`

hcmh commented:
So, I adapted the coding style, fixed the start positions for negative `dx`/`dy`, and replaced the `error()` with `assert(0)`.

The computation of the start can indeed be simplified, but I don't see a way of completely removing the condition: For `dx`/`dy == 0`, I want the corresponding shift to also equal 0, which I can only get through a comparison. But I have written it in a way which avoids the if/else and is still readable, in my opinion.

uecker commented:
Ok, looks good! Just two more little things: But can you drop the <stdio.h> patch and add my little grafic to the comment.

hcmh commented:
Done.
